src/driver/AmdCompiler.cpp

```diff
@@ -772,6 +773,19 @@ bool AMDGPUCompiler::ParseLLVMOptions(const std::vector<std::string>& options) {
   return true;
 }
 
+bool AMDGPUCompiler::LoadPlugins(const std::vector<std::string>& plugins) {
+  if (plugins.empty()) { return true; }
+  // Load any requested plugins.
+  for (const auto &path : plugins) {
+    std::string error;
+    if (llvm::sys::DynamicLibrary::LoadLibraryPermanently(path.c_str(), &error)) {
```

Comment:
It looks like LLVM is able only to load library permanently and not to unload: 

[Note: there is currently no interface for temporarily loading a library](https://github.com/llvm-mirror/llvm/blob/master/include/llvm/Support/DynamicLibrary.h)

But the in-process mode of the compiler implies that it is always in memory and might live as long as the Runtime by which it's invoked. It also means that compilation might be called (and is called) multiple times for different sources and with different options, including options for loading plugins. I believe that we should load plugins per one compilation (clang::driverCompilation class instance).

So we need a solution for unloading plugin library. What do you think?

Reply:
Well, I'm unsure if there is a specific reason that there is no interface for temporarily loading a library. It looks like libraries are unloaded when [llvm::llvm_shutdown](https://github.com/llvm-mirror/llvm/blob/master/lib/Support/ManagedStatic.cpp#L80) is called. AFAICS with AMDOCL in-process llvm_shutdown is never called. We cannot call llvm_shutdown after each compilation because it will effect more than just loaded libraries. So I guess one option would be to re-initialize LLVM before each compilation so that llvm_shutdown can be called.

I don't think there is a good solution for unloading a plugin library without patching LLVM. I tried something like this:
```
void DynamicLibrary::UnloadLibraries() {
  SmartScopedLock<true> Lock(*SymbolsMutex);
  // 
  // We would also need to Lock ManagedStaticMutex
  // here I guess. 
  //
  // Unload dynamic libraries by destroying HandleSet.
  if (OpenedHandles.isConstructed())
    OpenedHandles.destroy();
}
```

The call to [::dlclose](https://github.com/llvm-mirror/llvm/blob/master/lib/Support/Unix/DynamicLibrary.inc#L20) doesn't necessarily unload the library, because it seems like I can still use the plugin for the next clang::driverCompilation even without reloading it.

Reply:
What about switching off in-process in case of loading plugin(s) (as long as it seems there is no solution for temporary loading plugins from libraries, at least yet)? 

Btw, llvm_shutdown needs LLVM (re)initialization, but it also doesn't work since llvm_shutdown was designed to be called just before application termination and actually does not imply further work with LLVM.

Reply:
So, let's leave the change with in-process for a while, and make instead switching in-process off in case of loading plugin (corresponding options are found) and switching in-process back on after finishing compilation. In case of in-process off do nothing.